src/gadget/monotone.rs

```diff
@@ -0,0 +1,259 @@
+use halo2::{
+    circuit::{Chip, Layouter},
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Fixed, VirtualCells},
+    poly::Rotation,
+};
+use pasta_curves::arithmetic::FieldExt;
+use std::{marker::PhantomData, u64};
+
+#[derive(Clone, Debug)]
+pub(crate) struct MonotoneConfig {
+    range_table: Column<Fixed>,
+}
+
+/// MonotoneChip helps to check if an advice column is monotone increasing
+/// within a range. With strict enabled, it disallows equality of two cell.
+pub(crate) struct MonotoneChip<F, const RANGE: usize, const INCR: bool, const STRICT: bool> {
+    config: MonotoneConfig,
+    _marker: PhantomData<F>,
+}
+
+pub(crate) type StrictMonoIncrChip<F, const RANGE: usize> = MonotoneChip<F, RANGE, true, true>;
+pub(crate) type NonStrictMonoIncrChip<F, const RANGE: usize> = MonotoneChip<F, RANGE, true, false>;
+
+impl<F: FieldExt, const RANGE: usize, const INCR: bool, const STRICT: bool>
+    MonotoneChip<F, RANGE, INCR, STRICT>
+{
+    /// configure which column should be check. q_enable here as a fn is
+    /// flexible for synthetic selector instead of a fixed one.
+    pub fn configure(
+        meta: &mut ConstraintSystem<F>,
+        q_enable: impl FnOnce(&mut VirtualCells<'_, F>) -> Expression<F>,
```

Comment:
It may be simpler here to just pass in `q_enable: Expression<F>`, which is still general enough for a synthetic selector.

Reply:
I tried to pass `Expression<F>` at first actually, but the only way to create a `Expression` and put the column into `Gate.queried_selectors` is to call `VirtualCells`'s function, and the only way we have `VirtualCells` is to borrow `mut ConstraintSystem` (call `create_gate` or `lookup`), but the first parameter of `configure` is also a borrowed `mut ConstraintSystem`, which will be blocked by borrow checker. So finally I yield to pass a function to get the expression without borrowing `mut ConstraintSystem`.

Maybe there is other better way to pass through this, do you have any suggestion I can try?

Reply:
Let's use this pattern for now!